pkg/blocks/step.go

```diff
@@ -22,241 +22,176 @@ package blocks
-	WorkDir     string                  `yaml:"-"`
-	Outputs     map[string]outputs.Spec `yaml:"outputs,omitempty"`
-	Type        StepType                `yaml:"-"`
+	// CleanupSpec is exported so that UnmarshalYAML
+	// can see it - however, it should be considered
+	// to be a private detail of this file
+	// and not referenced elsewhere in the codebase
+	CleanupSpec yaml.Node `yaml:"cleanup,omitempty"`
 }
 
-// CleanupAct interface is implemented by anything that requires a cleanup step.
-type CleanupAct interface {
-	Cleanup(execCtx TTPExecutionContext) (*ActResult, error)
-	StepName() string
-	SetDir(dir string)
-	IsNil() bool
-	Validate(execCtx TTPExecutionContext) error
+// Step contains a TTPForge executable action
+// and its associated cleanup action (if specified)
+type Step struct {
+	CommonStepFields
+
+	// These are where the actual executable content
+	// of the step (and its associated cleanup process)
+	// live - they are not deserialized directly from YAML
+	// but rather must be decoded by ParseAction
+	action  Action
+	cleanup Action
 }
 
-// Step is an interface that represents a TTP step. Types that implement
-// this interface must provide methods for setting up the environment and
-// output references, setting the working directory, getting the cleanup
-// actions, executing the step, checking if the step is empty, explaining
-// validation errors, validating the step, fetching arguments, getting output,
-// searching output, setting output success status, checking success status,
-// returning the step name, and getting the step type.
-type Step interface {
-	SetDir(dir string)
-	// Need list in case some steps are encapsulating many cleanup steps
-	GetCleanup() []CleanupAct
-	// Execute will need to take care of the condition checks/etc...
-	Execute(execCtx TTPExecutionContext) (*ExecutionResult, error)
-	IsNil() bool
-	ExplainInvalid() error
-	Validate(execCtx TTPExecutionContext) error
-	StepName() string
-	GetType() StepType
-}
+func isDefaultCleanup(cleanupNode *yaml.Node) (bool, error) {
+	var testStr string
+	// is it a string? if not, let the subsequent decoding
+	// in the calling function deal with it
+	if err := cleanupNode.Decode(&testStr); err != nil {
+		return false, nil
+	}
 
-// SetDir sets the working directory for the Act.
-//
-// **Parameters:**
-//
-// dir: A string representing the directory path to be set
-// as the working directory.
-func (a *Act) SetDir(dir string) {
-	a.WorkDir = dir
+	// if it is a string, it must be a valid string
+	if testStr == "default" {
+		return true, nil
+	}
+	return false, fmt.Errorf("invalid cleanup value specified: %v", testStr)
 }
 
-// IsNil checks whether the Act is nil (i.e., it does not have a name).
-//
-// **Returns:**
-//
-// bool: True if the Act has no name, false otherwise.
-func (a *Act) IsNil() bool {
-	switch {
-	case a.Name == "":
+// ShouldCleanupOnFailure specifies that this step should be cleaned
+// up even if its Execute(...)  failed.
```

Comment:
these two functions are hacks to force the behavior of "cleanup is run even on failure, and default cleanup is run even when `cleanup: default` was not specified" specifically for SubTTPs - this is necessary for backward compatibility